### PR TITLE
[SPARK-44109][CORE] Remove duplicate preferred locations of each RDD partition

### PR DESCRIPTION
**Pull request by wankunde**
### What changes were proposed in this pull request?

DAGScheduler will get the preferred locations for each RDD partition and try to allocate the task on the preferred locations.

We can remove the duplicate preferred locations to save memory.

For example. reduce 0 needs to fetch map0 output and map1 output in host-A, then the preferred locations can be Array("host-A").

### Why are the changes needed?

To save driver memory for large job, too many preferred locations may cause OOM.

```
2023-06-19T14:14:18.501-0700: [Full GC (Allocation Failure) 2023-06-19T14:14:18.501-0700: [Class Histogram (before full gc):
 num     #instances         #bytes  class name
----------------------------------------------
   1:    2980686820    95381978240  scala.collection.immutable.$colon$colon
   2:    2978345818    71480299632  org.apache.spark.scheduler.HostTaskLocation
   3:      24388384     3963604560  [C
   4:      24265128      776484096  java.lang.String
   5:        368944      560810384  [Ljava.lang.Object;
   6:       2927846      421609824  java.net.URI
   7:      16239585      389750040  java.lang.Integer
   8:       2540273      243866208  org.apache.hadoop.fs.FileStatus
   9:        676141      203208736  [B
  10:       3845164      184567872  java.util.Hashtable$Entry
  11:       3174033      152353584  java.util.concurrent.ConcurrentHashMap$Node
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added UT

### CODE REVIEW

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @wankunde .

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 1]
```diff
@@ -1122,7 +1122,7 @@ private[spark] class MapOutputTrackerMaster(
         if (startMapIndex < endMapIndex &&
           (startMapIndex >= 0 && endMapIndex <= statuses.length)) {
           val statusesPicked = statuses.slice(startMapIndex, endMapIndex).filter(_ != null)
-          statusesPicked.map(_.location.host).toSeq
+          statusesPicked.map(_.location.host).toSeq.distinct
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala:1125]: +1, for @LuciferYang 's comment.

[the remark this answers is not in the document]
Comment (wankunde) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala]: Thanks.

**Review by dongjoon-hyun — APPROVED** [review]
Thank you for updating. Pending CIs.

**Review by wangyum — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Hi, @wankunde . Could you take a look at your GitHub Action pipeline? One job seems to be stuck.

**Comment by wankunde** [comment]
Retest this please

**Review by LuciferYang — APPROVED** [review]
+1, LGTM, pending CIs

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @wankunde , @LuciferYang and @wangyum .

**Comment by mridulm** [comment]
Not handling this for shuffle ?

**Comment by mridulm** [comment]
> Hi, @mridulm Thanks for your reminder. Can you describe the non-shuffle scenario more specifically?

It is not very common for shuffle stages to have preferred locality (unless push based - where node is always 1 or nontrivial fraction of reducer input generated in an executor).
It is much more common for non-shuffle stages to surface locality.
Depending on the underlying impl, this can or need not have duplication: hdfs for example does not have duplication, while replicated RDD could, and so on.

**Pull request closed without merging**